src/inc/cor.h

```diff
@@ -2134,6 +2134,15 @@ inline ULONG CorSigUncompressData(      // return number of bytes of that compre
 #endif
 #endif
 
+#if !defined(SELECTANY_NOEXTERN)
+#if defined(__GNUC__)
+    #define SELECTANY_NOEXTERN __attribute__((weak))
```

Comment:
Why is it important to omit the `extern` ?

Reply:
I tried to capture this in the commit message.

src/inc/clrproviders.h:526:32: error: ‘MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_DOTNET_Context’ initialized and declared ‘extern’ [-Werror]

We can have selectany and extern const variable.
We can't have selectany and extern variable.

There is a Microsoft page about legal and illegal use cases.

https://docs.microsoft.com/en-us/cpp/cpp/selectany?view=vs-2019

Reply:
Would it be an option to mark these as `const` instead?

Reply:
Sure, I can try that.

Reply:
it didn't quiet work.

```
src/inc/clrproviders.h:528:235: error: invalid conversion from ‘const LTTNG_TRACE_CONTEXT* {aka const _LTTNG_TRACE_CONTEXT*}’ to ‘PLTTNG_TRACE_CONTEXT {aka _LTTNG_TRACE_CONTEXT*}’ [-fpermissive]
 SELECTANY const DOTNET_TRACE_CONTEXT MICROSOFT_WINDOWS_DOTNETRUNTIME_STRESS_PROVIDER_DOTNET_Context = { MICROSOFT_WINDOWS_DOTNETRUNTIME_STRESS_PROVIDER_EVENTPIPE_Context, &MICROSOFT_WINDOWS_DOTNETRUNTIME_STRESS_PROVIDER_LTTNG_Context };
```

Reply:
Is this field written to anywhere? Can we mark it as const?

Reply:
No, it is not written anywhere. If you look at the example, we are having a hard time initializing a const weak structure.

I tried different combinations like moving the const around. No luck. I also must admit that I'm not the best skilled person when it comes to const.

Reply:
I managed to make them const but they are being modified here:

    static void ActivateAllKeywordsOfAllProviders()
    {
        for (const LTTNG_TRACE_CONTEXT * const provider : ALL_LTTNG_PROVIDERS_CONTEXT)
        {
            provider->EnabledKeywordsBitmask = (ULONGLONG)(-1);
            provider->Level = TRACE_LEVEL_VERBOSE;
            provider->IsEnabled = true;
        }
    }
    
    src/inc/eventtracebase.h
    

coreclr/src/inc/eventtracebase.h:398:46: error: assignment of member ‘_LTTNG_TRACE_CONTEXT::EnabledKeywordsBitmask’ in read-only object
             provider->EnabledKeywordsBitmask = (ULONGLONG)(-1);
                                              ^
coreclr/src/inc/eventtracebase.h:399:29: error: assignment of member ‘_LTTNG_TRACE_CONTEXT::Level’ in read-only object
             provider->Level = TRACE_LEVEL_VERBOSE;
                             ^
coreclr/src/inc/eventtracebase.h:400:33: error: assignment of member ‘_LTTNG_TRACE_CONTEXT::IsEnabled’ in read-only object
             provider->IsEnabled = true;
                                 ^
/

Reply:
coreclr/src/vm/eventpipesession.cpp:412:93: error: assignment of member ‘_EVENTPIPE_TRACE_CONTEXT::Level’ in read-only object
     MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_DOTNET_Context.EventPipeProvider.Level = VerboseLoggingLevel;
                                                                                             ^
coreclr/src/vm/eventpipesession.cpp:413:110: error: assignment of member ‘_EVENTPIPE_TRACE_CONTEXT::EnabledKeywordsBitmask’ in read-only object
     MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_DOTNET_Context.EventPipeProvider.EnabledKeywordsBitmask = Keywords;
                                                                                                              ^
coreclr/src/vm/eventpipesession.cpp:414:97: error: assignment of member ‘_EVENTPIPE_TRACE_CONTEXT::IsEnabled’ in read-only object
     MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_DOTNET_Context.EventPipeProvider.IsEnabled = true;

Reply:
@jkotas : how do we proceed?

Reply:
I think it would be best to refactor this to not use SELECTANY. 

Non-extern SELECTANY statics that are written to are recipe for problems. I believe that the compiler is free to create more copies of such statics and the write is going to modify just one of them.

@noahfalk Who would be the best person on diagnostic team to provide guidance with this change?

In any case, we should verify that the change is not introducing a regression. The default CI runs do not have automated testing for LTTNG tracing that this is changing.

Reply:
@sywhang is probably the right person. We should be able to avoid the selectany, but it may take some more messaging of the auto-generated code to get there.

Reply:
@sywhang: can you help us here?

Reply:
Sorry, this slipped through my inbox. I'm running tests on this change now.